File: backend/src/engines/ViolationEngine.ts
import type { FineRule, ViolationEstimate } from '../types.js'

export const ESTIMATE_DISCLAIMER =
  'Estimativa apenas — possível infração. Não constitui autuação oficial.'

/**
 * Motor de classificação CTB art. 218.
 * Lê regras de fine_rules; NUNCA hardcoda valores de multa.
 */
export class ViolationEngine {
  constructor(private rules: FineRule[]) {}

  setRules(rules: FineRule[]): void {
    this.rules = rules.filter((r) => r.active)
  }

  /** Excesso percentual sobre o limite. */
  excessPct(speedKmh: number, limitKmh: number): number {
    if (limitKmh <= 0) return 0
    const excess = Math.max(0, speedKmh - limitKmh)
    return (excess / limitKmh) * 100
  }

  excessKmh(speedKmh: number, limitKmh: number): number {
    return Math.max(0, speedKmh - limitKmh)
  }

  /**
   * Localiza a regra ativa cujo intervalo cobre o excesso %.
   * Faixas: (0, 20], (20, 50], (50, ∞) alinhadas ao CTB art. 218 seed.
   */
  lookupRule(excessPct: number): FineRule | null {
    if (excessPct <= 0) return null
    const active = this.rules
      .filter((r) => r.active)
      .sort((a, b) => a.excessMinPct - b.excessMinPct)

    for (const rule of active) {
      const lower =
        rule.excessMinPct === 0
          ? excessPct > 0
          : excessPct > rule.excessMinPct
      const upper =
        rule.excessMaxPct == null ? true : excessPct <= rule.excessMaxPct
      if (lower && upper) return rule
    }
    return null
  }

  classify(speedKmh: number, limitKmh: number): ViolationEstimate {
    const excessKmh = this.excessKmh(speedKmh, limitKmh)
    const excessPct = this.excessPct(speedKmh, limitKmh)

    if (excessKmh <= 0 || limitKmh <= 0) {
      return {
        excessKmh: 0,
        excessPct: 0,
        severity: null,
        fineRuleCode: null,
        estimatedFineBrl: null,
        points: null,
        suspendRight: false,
        disclaimer: ESTIMATE_DISCLAIMER,
        label: 'Dentro do limite (estimativa)',
      }
    }

    const rule = this.lookupRule(excessPct)
    if (!rule) {
      return {
        excessKmh,
        excessPct,
        severity: null,
        fineRuleCode: null,
        estimatedFineBrl: null,
        points: null,
        suspendRight: false,
        disclaimer: ESTIMATE_DISCLAIMER,
        label: 'Possível infração sem regra configurada',
      }
    }

    return {
      excessKmh,
      excessPct,
      severity: rule.severity,
      fineRuleCode: rule.code,
      estimatedFineBrl: rule.fineBrl,
      points: rule.points,
      suspendRight: rule.suspendRight,
      disclaimer: ESTIMATE_DISCLAIMER,
      label: `Estimativa de possível infração (${rule.severity})`,
    }
  }
}

File: ios/RelatorioRadar/Sources/RelatorioRadarCore/Engines/ViolationEngine.swift
import Foundation

/// Avalia excesso vs limite usando faixas CTB 218.
/// Valores monetários **somente** via `FineRule` injetadas — nunca hardcoded.
public final class ViolationEngine: @unchecked Sendable {
    public var fineRules: [FineRule]
    public var minConfidenceForRed: Double

    public init(fineRules: [FineRule] = [], minConfidenceForRed: Double = 0.65) {
        self.fineRules = fineRules
        self.minConfidenceForRed = minConfidenceForRed
    }

    /// Classifica excesso percentual (CTB art. 218).
    public static func severity(forExcessPercent percent: Double) -> ViolationSeverity {
        if percent <= 0 { return .none }
        if percent <= 20 { return .medium }
        if percent <= 50 { return .serious }
        return .verySerious
    }

    public func evaluate(passage: RadarPassage) -> ViolationEstimate {
        let limit = max(1, passage.radar.limitKmh)
        let speed = max(0, passage.speedKmh)
        let excess = max(0, speed - limit)
        let percent = (excess / limit) * 100
        let severity = Self.severity(forExcessPercent: percent)
        let confidence = passage.confidence

        let marker = markerColor(
            severity: severity,
            confidence: confidence,
            excessPercent: percent
        )

        let rule = fineRules.first { $0.severity == severity && severity != .none }

        let note: String
        switch marker {
        case .gray:
            note = "Estimativa indisponível: baixa confiança do GPS/matching."
        case .yellow:
            note = "Possível excesso (estimativa). Confirme com dados oficiais."
        case .red:
            note = "Excesso estimado com alta confiança. Não é multa oficial."
        case .green:
            note = "Dentro do limite estimado."
        }

        return ViolationEstimate(
            passageId: passage.id,
            radarId: passage.radar.id,
            radarName: passage.radar.name,
            limitKmh: limit,
            measuredSpeedKmh: speed,
            excessKmh: excess,
            excessPercent: percent,
            severity: severity,
            markerColor: marker,
            confidence: confidence,
            estimatedFineBRL: rule?.amountBRL,
            estimatedPoints: rule?.points,
            fineRuleId: rule?.id,
            isEstimate: true,
            notePT: note,
            occurredAt: passage.passedAt
        )
    }

    public func evaluate(passages: [RadarPassage]) -> [ViolationEstimate] {
        passages.map { evaluate(passage: $0) }
    }

    public func markerColor(
        severity: ViolationSeverity,
        confidence: Double,
        excessPercent: Double
    ) -> ViolationMarkerColor {
        if severity == .none {
            return confidence < 0.35 ? .gray : .green
        }
        // Gates de confiança: baixa → gray; média → yellow; alta → red.
        if confidence < 0.4 {
            return .gray
        }
        if confidence < minConfidenceForRed || excessPercent < 3 {
            return .yellow
        }
        return .red
    }
}
